internal/storage/postgres_storage.go

```diff
@@ -0,0 +1,414 @@
+	query := `SELECT user_id FROM orders WHERE order_number = $1`
+	r, err := p.dbConn.Query(opCtx, query, orderNumber)
+	if err != nil {
+		return err
+	}
+
+	if err := r.Err(); err != nil {
+		return err
+	}
+	defer r.Close()
+
+	if r.Next() {
+		clientID := uuid.UUID{}
+		if err := r.Scan(&clientID); err != nil {
+			return err
+		}
+		if clientID == userID {
+			return ErrOrderAlreadyPlaced
+		}
+	}
+	return ErrDuplicateOrder
+}
+
+func (p *pgxStorage) UpdateOrder(ctx context.Context, order Order) error {
+	opCtx, cancel := context.WithTimeout(ctx, DatabaseOperationTimeout)
+	defer cancel()
+
+	tx, err := p.dbConn.Begin(opCtx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(p.ctx)
+
+	p.logger.Info("updating order", zap.Any("order_number", order.OrderNumber), zap.Float64("accrual", order.Accrual))
+	_, err = tx.Exec(opCtx, `UPDATE orders SET status=$1, accrual=$2, updated_at=NOW() WHERE order_number=$3;`, order.Status, order.Accrual, order.OrderNumber)
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit(opCtx)
+}
+
+func (p *pgxStorage) GetOrders(ctx context.Context, userID uuid.UUID) ([]Order, error) {
+	opCtx, cancel := context.WithTimeout(ctx, DatabaseOperationTimeout)
+	defer cancel()
+
+	r, err := p.dbConn.Query(opCtx, `SELECT order_number, status, accrual, uploaded_at FROM orders WHERE user_id = $1;`, userID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Err(); err != nil {
+		return nil, err
+	}
+
+	defer r.Close()
+
+	orders := make([]Order, 0)
+	for r.Next() {
+		order := Order{
+			UserID: userID,
+		}
+		if err := r.Scan(&order.OrderNumber, &order.Status, &order.Accrual, &order.UploadedAt); err != nil {
+			return nil, err
+		}
+		orders = append(orders, order)
+	}
+
+	return orders, nil
+}
+
+func (p *pgxStorage) GetUnfinishedOrders(ctx context.Context) ([]Order, error) {
+	opCtx, cancel := context.WithTimeout(ctx, DatabaseOperationTimeout)
+	defer cancel()
+
+	r, err := p.dbConn.Query(opCtx, `SELECT order_number, user_id, status, accrual, uploaded_at FROM orders WHERE status IN ('NEW', 'PROCESSING');`)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Err(); err != nil {
+		return nil, err
+	}
+
+	defer r.Close()
+
+	orders := make([]Order, 0)
+	for r.Next() {
+		order := Order{}
+		var userID uuid.UUID
+		if err := r.Scan(&order.OrderNumber, &userID, &order.Status, &order.Accrual, &order.UploadedAt); err != nil {
+			return nil, err
+		}
+		order.UserID = userID
+		orders = append(orders, order)
+	}
+
+	p.logger.Info("unfinished orders", zap.Any("orders", orders))
+
+	return orders, nil
+}
+
+func (p *pgxStorage) Withdraw(ctx context.Context, userID uuid.UUID, order string, sum float64) error {
+	opCtx, cancel := context.WithTimeout(ctx, DatabaseOperationTimeout)
+	defer cancel()
+
+	tx, err := p.dbConn.Begin(opCtx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(p.ctx)
+
+	r, err := tx.Query(opCtx, `SELECT current, withdrawn FROM balance WHERE user_id = $1;`, userID)
```

Comment:
```suggestion
	r, err := tx.Query(opCtx, `SELECT current, withdrawn FROM balance WHERE user_id = $1 FOR UPDATE;`, userID)
```
https://t.me/bushigo/36

Reply:
посмотрел, пост из тг. добавил `FOR UPDATE`